### DISCUSSION TITLE
Connection object released back to the pool after commit()

### DISCUSSION THREAD

**Discussion by joan1294** [sqlalchemy/sqlalchemy#11571] [Usage Questions] [answered]
### Describe the bug

I am using sqlalchemy to manage an sql server database from fastapi. For each endpoint execution, I need to set up a context variable to the session with `session.execute("sys.sp_set_session_context 'context_var', 'value'")`  which will be used for row-level security.

I found out that when I execute multiple commits within an endpoint function, there is a chance that when executing session.commit(), the connection object is released to the Pool and a new one is acquired which will be used in the following commits. This forces me to reset the context variable every time after each `session.commit()` to maintain the correct row-level security. The closest case related to my issue I could find is https://github.com/sqlalchemy/sqlalchemy/issues/4925, but in my case I don't think I am deleting any strong reference as mentioned in:
>  when you create a connection pool, then acquire some connections, and then you delete all strong references to those connections, the connection pool uses a weakref callback that still contains a strong reference to pool's internal record of this connection as well as the DBAPI connection itself, so that it can be brought back into the pool as though it were explicitly returned

Is there a way to keep the same connection object consistently until the `session.close()`?. Is there a better way to manage this? I would also like to avoid solutions such as using a `NullPool` to prevent jumping from one connection to another.

Commiting just once per endpoint function is not an option for me since I am doing some other stuff in between commits which require these separate commits.

### SQLAlchemy Version in Use

1.4.42

### DBAPI (i.e. the database driver)

pyodbc

### Database Vendor and Major Version

SQL Server 2022

### Python Version

3.11

### Operating system

Linux

### To Reproduce

```python
import random
from sqlalchemy.orm import Session

engine = create_engine(
        "cnn_string",
        future=True,
        implicit_returning=False,
        pool_pre_ping=True,
        pool_recycle=3600,
        fast_executemany=True,
    )]
SessionMaker = sessionmaker(bind=engine, future=True)

def get_session():
    db = SessionMaker()
    try:
        db.execute(f"sys.sp_set_session_context 'N'context_var', '{random.choice([0, 1, 2])}'")
        yield db
    finally:
        db.close()

@router.post("/create-users")
async def endpoint1(db: Session = Depends(get_session)):
   user = User(name="user1")
   db.add(object1)
   value_before_commit = db.execute("SELECT CAST(SESSION_CONTEXT(N'context_var') AS VARCHAR)").all()[0][0]
   print(value_before_commit)
   db.commit()
   value_after_commit = db.execute("SELECT CAST(SESSION_CONTEXT(N'context_var') AS VARCHAR)").all()[0][0]
   print(value_after_commit)
   assert value_before_commit == value_after_commit, "Connection object changed"
   
   db.add(object2)
   db.commit()
```

### Error

```python
AssertionError
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

That's expected. The session will release the connection to the pool on commit when the bind is an engine. To avoid that bind the session on a connection directly. Something like the following, but I don't know what AdminSession is

```py
def get_session():
    with engine.connect() as conn:
      db = AdminSession(bind=conn, info={"token_hash": token_hash_value})
      try:
          db.execute(f"sys.sp_set_session_context 'TokenHash', '{token_hash_value}'")
          yield db
      finally:
          db.close()
